src/azure-cli-core/azure/cli/core/credential.py

```diff
@@ -0,0 +1,127 @@
+        except CLIError as err:
+            if in_cloud_console():
+                CredentialAdaptor._log_hostname()
+            raise err
+        except ClientAuthenticationError as err:
+            # pylint: disable=no-member
+            if in_cloud_console():
+                CredentialAdaptor._log_hostname()
+
+            err = getattr(err, 'message', None) or ''
+            if 'authentication is required' in err:
+                raise CLIError("Authentication is migrated to Microsoft identity platform (v2.0). {}".format(
+                    "Please run 'az login' to login." if not in_cloud_console() else ''))
+            if 'AADSTS70008' in err:  # all errors starting with 70008 should be creds expiration related
+                raise CLIError("Credentials have expired due to inactivity. {}".format(
+                    "Please run 'az login'" if not in_cloud_console() else ''))
+            if 'AADSTS50079' in err:
+                raise CLIError("Configuration of your account was changed. {}".format(
+                    "Please run 'az login'" if not in_cloud_console() else ''))
+            if 'AADSTS50173' in err:
+                raise CLIError("The credential data used by CLI has been expired because you might have changed or "
+                               "reset the password. {}".format(
+                                   "Please clear browser's cookies and run 'az login'"
+                                   if not in_cloud_console() else ''))
+            raise CLIError(err)
+        except requests.exceptions.SSLError as err:
+            from .util import SSLERROR_TEMPLATE
+            raise CLIError(SSLERROR_TEMPLATE.format(str(err)))
+        except requests.exceptions.ConnectionError as err:
+            raise CLIError('Please ensure you have network connection. Error detail: ' + str(err))
+        return token, external_tenant_tokens
+
+    def signed_session(self, session=None):
+        logger.debug("CredentialAdaptor.signed_session invoked by Track 1 SDK")
+        session = session or requests.Session()
+        token, external_tenant_tokens = self._get_token()
+        header = "{} {}".format('Bearer', token.token)
+        session.headers['Authorization'] = header
+        if external_tenant_tokens:
+            aux_tokens = ';'.join(['{} {}'.format('Bearer', tokens2.token) for tokens2 in external_tenant_tokens])
+            session.headers['x-ms-authorization-auxiliary'] = aux_tokens
+        return session
+
+    def get_token(self, *scopes):
+        # type: (*str) -> AccessToken
+        logger.debug("CredentialAdaptor.get_token invoked by Track 2 SDK with scopes=%r", scopes)
+        scopes = _normalize_scopes(scopes)
+        token, _ = self._get_token(scopes)
+        return token
+
+    def get_all_tokens(self, *scopes):
+        # type: (*str) -> Tuple[AccessToken, List[AccessToken]]
+        # TODO: Track 2 SDK should support external credentials.
+        return self._get_token(scopes)
```

Comment:
This function is added by #15750 to workaround the "Track 2 SDK doesn't support external credential" issue (https://github.com/Azure/azure-sdk-for-python/issues/8313) by directly attaching the token as `x-ms-authorization-auxiliary` header:

https://github.com/Azure/azure-cli/blob/5aa3d4e45f4894efe48a6b2606c32993e21cef96/src/azure-cli/azure/cli/command_modules/vm/custom.py#L3273-L3276

https://github.com/Azure/azure-cli/blob/5aa3d4e45f4894efe48a6b2606c32993e21cef96/src/azure-cli/azure/cli/command_modules/vm/custom.py#L3329-L3336

Directly manipulating the token is not correct in **Track 2 + Azure Identity** workflow. This function should be removed and Track 2 SDK should support external credential instead.

Reply:
What's the current solution when no Track 2 SDK external token support?

Reply:
Your code is indeed the only solution, but it is against the Track 2 design.